src/RadixDlt.NetworkGateway.PostgresIntegration/Interceptors/MetricsInterceptor.cs

```diff
@@ -0,0 +1,99 @@
+﻿using Microsoft.EntityFrameworkCore.Diagnostics;
+using Microsoft.EntityFrameworkCore.Storage.Internal;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using RadixDlt.NetworkGateway.GatewayApi.Configuration;
+using RadixDlt.NetworkGateway.GatewayApi.Services;
+using RadixDlt.NetworkGateway.PostgresIntegration.Metrics;
+using System.Data.Common;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace RadixDlt.NetworkGateway.PostgresIntegration.Interceptors;
+
+internal class MetricsInterceptor : DbCommandInterceptor
+{
+    private const string UnknownQueryName = "UNKNOWN";
+    private readonly IOptionsMonitor<SlowQueriesLoggingOptions> _slowQueriesLoggingOptions;
+    private readonly ILogger<MetricsInterceptor> _logger;
+    private readonly ISqlQueryObserver _sqlQueryObserver;
+
+    public MetricsInterceptor(ILoggerFactory loggerFactory, IOptionsMonitor<SlowQueriesLoggingOptions> slowQueriesLoggingOptions, ISqlQueryObserver sqlQueryObserver)
+    {
+        _slowQueriesLoggingOptions = slowQueriesLoggingOptions;
+        _sqlQueryObserver = sqlQueryObserver;
+        _logger = loggerFactory.CreateLogger<MetricsInterceptor>();
+    }
+
+    public override DbDataReader ReaderExecuted(DbCommand command, CommandExecutedEventData eventData, DbDataReader result)
+    {
+        Observe(command, eventData);
+        return result;
+    }
+
+    public override ValueTask<DbDataReader> ReaderExecutedAsync(
+        DbCommand command,
+        CommandExecutedEventData eventData,
+        DbDataReader result,
+        CancellationToken cancellationToken = default)
+    {
+        Observe(command, eventData);
+        return new ValueTask<DbDataReader>(result);
+    }
+
+    private void Observe(
+        DbCommand command,
+        CommandExecutedEventData eventData)
+    {
+        var queryName = GetQueryName(command) ?? UnknownQueryName;
+
+        _sqlQueryObserver.OnSqlQueryExecuted(queryName, eventData.Duration);
+
+        var logQueriesLongerThan = _slowQueriesLoggingOptions.CurrentValue.SlowQueriesThreshold;
+        if (eventData.Duration > logQueriesLongerThan)
+        {
+#pragma warning disable EF1001
+            var parameters = command.Parameters.FormatParameters(true);
+#pragma warning restore EF1001
+
+            _logger.LogWarning(
+                "Long running query: {query}, parameters: {queryParameters} duration: {queryDuration} milliseconds",
+                command.CommandText, parameters, eventData.Duration.TotalMilliseconds);
+        }
+    }
+
+    private string? GetQueryName(DbCommand dbCommand)
+    {
+        const string QueryNameGroup = "queryName";
+        var matches = Regex.Matches(dbCommand.CommandText, $"(?:{SqlQueryMetricsHelper.QueryNameTag}=)(?<{QueryNameGroup}>\\b.*?\\b);");
```

Comment:
Could we please use precompiled regex?

Reply:
Or actually... wouldn't it be even easier to:

1. Find the position of `-- xxx` prefix. Fast fail if missing.
2. Find the position of EOL past whatever was the position of above. Fast fail if somehow missing (exception?)
3. Query name would be the substring between position no. 1 + prefix length and position no. 2.

Reply:
ok aggreed that we want to check if tag name is alphanumeric when adding and use substring = -> ;.